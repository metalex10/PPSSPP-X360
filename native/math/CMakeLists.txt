set(SRCS
  lin/matrix4x4.cpp
  lin/vec3.cpp
  lin/quat.cpp
  lin/aabb.cpp
	curves.cpp
	math_util.cpp
  )

set(SRCS ${SRCS})

add_library(lin STATIC ${SRCS})

if(UNIX)
  add_definitions(-fPIC)
endif(UNIX)
