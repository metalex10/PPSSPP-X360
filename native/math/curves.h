#pragma once

#include <base/basictypes.h>

// Easy curve computation for fades etc.


// output range: [0.0, 1.0]
float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength);
float linearIn(int t, int fadeInLength);
float linearOut(int t, int fadeInLength);

// smooth operator [0, 1] -> [0, 1]
float ease(float val);
float ease(int t, int fadeLength);

// need a bouncy ease

// waveforms [0, 1]
float sawtooth(int t, int period);

// output range: -1.0 to 1.0
float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength);
