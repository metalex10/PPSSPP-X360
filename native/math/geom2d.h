#pragma once

#include <cmath>

struct Point {
	Point() {}
	Point(float x_, float y_) : x(x_), y(y_) {}

	float x;
	float y;

	float distanceTo(const Point &other) const {
		float dx = other.x - x, dy = other.y - y;
		return sqrtf(dx*dx + dy*dy);
	}

	/*
	FocusDirection directionTo(const Point &other) const {
		int angle = atan2f(other.y - y, other.x - x) / (2 * M_PI) - 0.125;

	}*/
};


// Resolved bounds on screen after layout.
struct Bounds {
	Bounds() : x(0), y(0), w(0), h(0) {}
	Bounds(float x_, float y_, float w_, float h_) : x(x_), y(y_), w(w_), h(h_) {}

	bool Contains(float px, float py) const {
		return (px >= x && py >= y && px < x + w && py < y + h);
	}

	bool Intersects(const Bounds &other) const {
		return !(x > other.x2() || x2() < other.x || y > other.y2() || y2() < other.y);
	}

	void Clip(const Bounds &clipTo) {
		if (x < clipTo.x) {
			w -= clipTo.x - x;
			x = clipTo.x;
		}
		if (y < clipTo.y) {
			h -= clipTo.y - y;
			y = clipTo.y;
		}
		if (x2() > clipTo.x2()) {
			w = clipTo.x2() - x;
		}
		if (y2() > clipTo.y2()) {
			h = clipTo.y2() - y;
		}
	}

	float x2() const { return x + w; }
	float y2() const { return y + h; }
	float centerX() const { return x + w * 0.5f; }
	float centerY() const { return y + h * 0.5f; }
	Point Center() const {
		return Point(centerX(), centerY());
	}

	float x;
	float y;
	float w;
	float h;
};


