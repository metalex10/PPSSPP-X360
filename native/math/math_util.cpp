#include "math/math_util.h"
#include <stdlib.h>

#if defined(__ARM_ARCH_7A__) && !defined(BLACKBERRY)

void EnableFZ()
{
	int x;
	asm(
		"fmrx %[result],FPSCR \r\n"
		"orr %[result],%[result],#16777216 \r\n"
		"fmxr FPSCR,%[result]"
		:[result] "=r" (x) : :
	);
	//printf("ARM FPSCR: %08x\n",x);
}
#else
void EnableFZ()
{
	// TODO
}
#endif
