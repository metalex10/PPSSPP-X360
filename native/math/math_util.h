#ifndef _MATHUTILS_H
#define _MATHUTILS_H

#include <cmath>
#include <cstring>

#include "base/basictypes.h"

inline float sqr(float f)	{return f*f;}
inline float sqr_signed(float f) {return f<0 ? -f*f : f*f;}

typedef unsigned short float16;

// This ain't a 1.5.10 float16, it's a stupid hack format where we chop 16 bits off a float.
// This choice is subject to change. Don't think I'm using this for anything at all now anyway.
// DEPRECATED
inline float16 FloatToFloat16(float x) {
	int ix;
	memcpy(&ix, &x, sizeof(float));
	return ix >> 16;
}

inline float Float16ToFloat(float16 ix) {
	float x;
	memcpy(&x, &ix, sizeof(float));
	return x;
}

inline uint32_t RoundUpToPowerOf2(uint32_t v) {
	v--;
	v |= v >> 1;
	v |= v >> 2;
	v |= v >> 4;
	v |= v >> 8;
	v |= v >> 16;
	v++;
	return v;
}

#define PI 3.141592653589793f
#ifndef M_PI
#define M_PI 3.141592653589793f
#endif

// The stuff in this file is from all over the web, esp. dspmusic.org. I think it's all public domain.
// In any case, very little of it is used anywhere at the moment.

// Calculate pseudo-random 32 bit number based on linear congruential method. 
void SetSeed(unsigned int seed);
unsigned int GenerateRandomNumber();
inline float GenerateRandomFloat01() {
	return (float)((double)GenerateRandomNumber() / 0xFFFFFFFF);
}
inline float GenerateRandomSignedFloat() {
	return (float)((double)GenerateRandomNumber() / 0x80000000) - 1.0f;
}


inline float GaussRand()
{
	float R1 = GenerateRandomFloat01();
	float R2 = GenerateRandomFloat01();

	float X = sqrtf(-2.0f * logf(R1)) * cosf(2.0f * PI * R2);
	if (X > 4.0f) X = 4.0f;
	if (X < -4.0f) X = -4.0f;
	return X;
}

// Accuracy unknown
inline double atan_fast(double x) {
	return (x / (1.0 + 0.28 * (x * x)));
}


// linear -> dB conversion
inline float lin2dB(float lin) {
	const float LOG_2_DB = 8.6858896380650365530225783783321f;	// 20 / ln( 10 )
	return logf(lin) * LOG_2_DB;
}

// dB -> linear conversion
inline float dB2lin(float dB) {
	const float DB_2_LOG = 0.11512925464970228420089957273422f;	// ln( 10 ) / 20
	return expf(dB * DB_2_LOG);
}

inline bool my_isinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	return f2u.u == 0x7f800000 ||
		f2u.u == 0xff800000;
}

inline bool my_isnan(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa
	return ((f2u.u & 0x7F800000) == 0x7F800000) && (f2u.u & 0x7FFFFF);
}

inline bool my_isnanorinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa, infs have zero mantissa. That is, we just ignore the mantissa here.
	return ((f2u.u & 0x7F800000) == 0x7F800000);
}

inline int is_even(float d) {
	float int_part;
	modff(d / 2.0f, &int_part);
	return 2.0f * int_part == d;
}

// Rounds *.5 to closest even number
inline double round_ieee_754(double d) {
	float i = floor(d);
	d -= i;
	if (d < 0.5f)
		return i;
	if (d > 0.5f)
		return i + 1.0f;
	if (is_even(i))
		return i;
	return i + 1.0f;
}

union FP32 {
	uint32_t u;
	float f;
};

struct FP16 {
	uint16_t u;
};

// magic code from ryg: http://fgiesen.wordpress.com/2012/03/28/half-to-float-done-quic/
// See also SSE2 version: https://gist.github.com/rygorous/2144712
inline FP32 half_to_float_fast5(FP16 h)
{
	static const FP32 magic = { (127 + (127 - 15)) << 23 };
	static const FP32 was_infnan = { (127 + 16) << 23 };
	FP32 o;
	o.u = (h.u & 0x7fff) << 13;     // exponent/mantissa bits
	o.f *= magic.f;                 // exponent adjust
	if (o.f >= was_infnan.f)        // make sure Inf/NaN survive
		o.u |= 255 << 23;
	o.u |= (h.u & 0x8000) << 16;    // sign bit
	return o;
}

inline float ExpandHalf(uint16_t half) {
	FP16 fp16;
	fp16.u = half;
	FP32 fp = half_to_float_fast5(fp16);
	return fp.f;
}

// More magic code: https://gist.github.com/rygorous/2156668
inline FP16 float_to_half_fast3(FP32 f)
{
	static const FP32 f32infty = { 255 << 23 };
	static const FP32 f16infty = { 31 << 23 };
	static const FP32 magic = { 15 << 23 };
	static const uint32_t sign_mask = 0x80000000u;
	static const uint32_t round_mask = ~0xfffu;
	FP16 o = { 0 };

	uint32_t sign = f.u & sign_mask;
	f.u ^= sign;

	if (f.u >= f32infty.u) // Inf or NaN (all exponent bits set)
		o.u = (f.u > f32infty.u) ? 0x7e00 : 0x7c00; // NaN->qNaN and Inf->Inf
	else // (De)normalized number or zero
	{
		f.u &= round_mask;
		f.f *= magic.f;
		f.u -= round_mask;
		if (f.u > f16infty.u) f.u = f16infty.u; // Clamp to signed infinity if overflowed

		o.u = f.u >> 13; // Take the bits!
	}

	o.u |= sign >> 16;
	return o;
}

inline uint16_t ShrinkToHalf(float full) {
	FP32 fp32;
	fp32.f = full;
	FP16 fp = float_to_half_fast3(fp32);
	return fp.u;
}

// FPU control.
void EnableFZ();

#endif
