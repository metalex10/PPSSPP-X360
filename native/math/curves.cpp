
#include <math.h>

#include "math/math_util.h"
#include "curves.h"

float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < solidLength) {
		return 1.0f;
	}
	t -= solidLength;
	if (t < fadeOutLength) {
		return 1.0f - (float)t / fadeOutLength;
	}
	return 0.0f;
}

float linearIn(int t, int fadeInLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	return 1.0f;
}

float linearOut(int t, int fadeOutLength) {
	return 1.0f - linearIn(t, fadeOutLength);
}

float ease(float val) {
	if (val > 1.0f) return 1.0f;
	if (val < 0.0f) return 0.0f;
	return ((-cosf(val * PI)) + 1.0f) * 0.5;
}

float ease(int t, int fadeLength)
{
	if (t < 0) return 0.0f;
	if (t >= fadeLength) return 1.0f;
	return ease((float)t / (float)fadeLength);
}

float sawtooth(int t, int period) {
	return (t % period) * (1.0f / (period - 1));
}

float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength)
{
	if (t < fadeInLength) {
		return -1.0f + (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < pauseLength) {
		return 0.0f;
	}
	t -= pauseLength;
	if (t < fadeOutLength) {
		return (float)t / fadeOutLength;
	}
	return 1.0f;
}
