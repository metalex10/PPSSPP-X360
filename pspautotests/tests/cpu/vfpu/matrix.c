/*
FPU Test. Originally from jpcsp project:
http://code.google.com/p/jpcsp/source/browse/trunk/demos/src/fputest/main.c
Modified to perform automated tests.
*/


// For some reason, vectors on the stack all are unaligned. So I've pulled them out
// into globals, which is horribly ugly. But it'll do for now.


#include <common.h>

#include <pspkernel.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>
//#include "../common/emits.h"

#include <pspgu.h>
#include <pspgum.h>

#include "vfpu_common.h"

__attribute__ ((aligned (16))) ScePspFVector4 v0, v1, v2, v3;
__attribute__ ((aligned (16))) ScePspFMatrix4 matrix;
__attribute__ ((aligned (16))) ScePspFMatrix4 matrix2;

void __attribute__((noinline)) vmidt(int size, ScePspFMatrix4 *v0, ScePspFMatrix4 *v1) {
	asm volatile (
		"lv.q    R000, %1\n"
		"lv.q    R001, %1\n"
		"lv.q    R002, %1\n"
		"lv.q    R003, %1\n"

		: "+m" (*v0) : "m" (*v1)
		);

	switch (size) {
	case 2: asm volatile("vmidt.p M000\nvmidt.p M022\n"); break;
	case 3: asm volatile("vmidt.t M000\n"); break;
	case 4: asm volatile("vmidt.q M000\n"); break;
	case 5: asm volatile("vmone.q M000\n"); break;
	case 6: asm volatile("vmzero.q M000\n"); break;

	}

	asm volatile (
		"sv.q    R000, 0x00+%0\n"
		"sv.q    R001, 0x10+%0\n"
		"sv.q    R002, 0x20+%0\n"
		"sv.q    R003, 0x30+%0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void checkMatrixIdentity() {
	int vsize;
	for (vsize = 2; vsize <= 6; vsize++) {
		nonsense(&matrix);
		vmidt(vsize, &matrix2, &matrix);
		printMatrix("vmidt", &matrix2);
	}
}

void _checkMultiply2(ScePspFMatrix4* m0, ScePspFMatrix4 *m1, ScePspFMatrix4 *m2) {
	__asm__ volatile (

		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		"lv.q    R100, 0x00+%2\n"
		"lv.q    R101, 0x10+%2\n"
		"lv.q    R102, 0x20+%2\n"
		"lv.q    R103, 0x30+%2\n"

		"vmmul.p M200, M000, M100\n"
		"vmmul.p M202, M002, E102\n"
		"vmmul.p E220, M020, M120\n"
		"vmmul.p M222, E022, M122\n"

		"sv.q    R200, 0x00+%0\n"
		"sv.q    R201, 0x10+%0\n"
		"sv.q    R202, 0x20+%0\n"
		"sv.q    R203, 0x30+%0\n"
		: "+m" (*m0) : "m" (*m1), "m" (*m2)
		);
}

void _checkMultiply4(ScePspFMatrix4* m0, ScePspFMatrix4 *m1, ScePspFMatrix4 *m2) {
	__asm__ volatile (

		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		"lv.q    R100, 0x00+%2\n"
		"lv.q    R101, 0x10+%2\n"
		"lv.q    R102, 0x20+%2\n"
		"lv.q    R103, 0x30+%2\n"

		"vmmul.q M200, M000, M100\n"

		"sv.q    R200, 0x00+%0\n"
		"sv.q    R201, 0x10+%0\n"
		"sv.q    R202, 0x20+%0\n"
		"sv.q    R203, 0x30+%0\n"
		: "+m" (*m0) : "m" (*m1), "m" (*m2)
		);
}

void _checkSquare4(ScePspFMatrix4* m0, ScePspFMatrix4 *m1) {
	__asm__ volatile (

		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		"vmmul.q M200, M000, M000\n"

		"sv.q    R200, 0x00+%0\n"
		"sv.q    R201, 0x10+%0\n"
		"sv.q    R202, 0x20+%0\n"
		"sv.q    R203, 0x30+%0\n"
		: "+m" (*m0) : "m" (*m1)
		);
}

void checkMultiply() {
	static ScePspFMatrix4 matrix1 = {
		{ 1.0f, 5.0f,  9.0f, 13.0f },
		{ 2.0f, 6.0f, 10.0f, 14.0f },
		{ 3.0f, 7.0f, 11.0f, 15.0f },
		{ 4.0f, 8.0f, 12.0f, 16.0f }
	};
	static ScePspFMatrix4 matrix2 = {
		{ -1.0f, 20.0f, 30.0f, -4.0f },
		{ 10.0f, 2.0f, 30000.0f, 400000.0f },
		{ 100.0f, 200.0f, -3.0f, 40.0f },
		{ -1000.0f, 2.0f, -0.3f, -400.0f }
	};

	_checkMultiply2(&matrix, &matrix1, &matrix2);
	printMatrix("vmmul.p", &matrix);
	_checkMultiply4(&matrix, &matrix1, &matrix2);
	printMatrix("vmmul.q", &matrix);
	_checkSquare4(&matrix, &matrix1);
	printMatrix("vmmul.q sq", &matrix);
}

static void doTranspose(ScePspFMatrix4* m0, ScePspFMatrix4 *m1) {
	__asm__ volatile (
		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		"sv.q    C000, 0x00+%0\n"
		"sv.q    C010, 0x10+%0\n"
		"sv.q    C020, 0x20+%0\n"
		"sv.q    C030, 0x30+%0\n"
		: "+m" (*m0) : "m" (*m1)
		);
}

void checkTranspose() {
	nonsense(&matrix2);
	printMatrix("non transpose", &matrix2);
	doTranspose(&matrix, &matrix2);
	printMatrix("transpose", &matrix);
}

void doMatrixScale(ScePspFMatrix4* m0, ScePspFMatrix4 *m1) {
	asm volatile (
		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		"vmscl.q M100, E000, S200\n"

		"sv.q    R100, 0x00+%0\n"
		"sv.q    R101, 0x10+%0\n"
		"sv.q    R102, 0x20+%0\n"
		"sv.q    R103, 0x30+%0\n"
		: "+m" (*m0) : "m" (*m1)
		);
}

void checkMatrixScale() {
	nonsense(&matrix);
	doMatrixScale(&matrix2, &matrix);
	printMatrix("vmscl.q", &matrix2);
}

void _checkMatrixPerVector(ScePspFMatrix4 *m, ScePspFVector4 *vmult, ScePspFVector4 *vresult) {
	asm volatile (
		"lv.q R700, 0x00+%1\n"
		"lv.q R701, 0x10+%1\n"
		"lv.q R702, 0x20+%1\n"
		"lv.q R703, 0x30+%1\n"

		"lv.q R600, 0x00+%2\n"

		"vtfm3.t R100, M700, R600\n"
		"vtfm4.q R200, M700, R600\n"

		"sv.q    R100, 0x00+%0\n"
		"sv.q    R200, 0x10+%0\n"
		: "+m" (*vresult) : "m" (*m), "m" (*vmult)
		);
}

void _checkHMatrixPerVector(ScePspFMatrix4 *matrix, ScePspFVector4 *vmult, ScePspFVector4 *vresult) {
	asm volatile (
		"lv.q R700, 0x00+%1\n"
		"lv.q R701, 0x10+%1\n"
		"lv.q R702, 0x20+%1\n"
		"lv.q R703, 0x30+%1\n"

		"lv.q R600, 0x00+%2\n"

		"vhtfm3.t R100, M700, R600\n"
		"vhtfm4.q R200, M700, R600\n"

		"sv.q    R100, 0x00+%0\n"
		"sv.q    R200, 0x10+%0\n"
		: "+m" (*vresult) : "m" (*matrix), "m" (*vmult)
		);
}


void checkMatrixPerVector() {
	static ScePspFVector4 vmult = { -13.0f, -2111.0f, 33.0f, 40.0f};
	static ScePspFVector4 vout[2];

	nonsense(&matrix);

	_checkMatrixPerVector(&matrix, &vmult, vout);
	printVector("vtfm3.t", &vout[0]);
	printVector("vtfm4.q", &vout[1]);

	_checkHMatrixPerVector(&matrix, &vmult, vout);
	printVector("vhtfm3.t", &vout[0]);
	printVector("vhtfm4.q", &vout[1]);
}

void checkMultiplyFull() {
	static __attribute__ ((aligned (16))) ScePspFMatrix4 m1 = {
		{  2,  3,  5,  7 },
		{ 11, 13, 17, 19 },
		{ 23, 29, 31, 37 },
		{ 41, 43, 47, 53 },
	};

	static __attribute__ ((aligned (16))) ScePspFMatrix4 m2 = {
		{  59,  61,  67,  71 },
		{  73,  79,  83,  89 },
		{  97, 101, 103, 107 },
		{ 109, 113, 127, 131 },
	};
	
	static __attribute__ ((aligned (16))) ScePspFMatrix4 m3;
	
	__attribute__ ((aligned (16))) ScePspFVector4 *v0 = NULL, *v1 = NULL;

	v1 = &m1.x;

	asm volatile (
		"lv.q    R000, 0x00+%1\n"
		"lv.q    R001, 0x10+%1\n"
		"lv.q    R002, 0x20+%1\n"
		"lv.q    R003, 0x30+%1\n"

		: "+m" (*v0) : "m" (*v1)
	);


	v1 = &m2.x;

	asm volatile (
		"lv.q    R100, 0x00+%1\n"
		"lv.q    R101, 0x10+%1\n"
		"lv.q    R102, 0x20+%1\n"
		"lv.q    R103, 0x30+%1\n"

		: "+m" (*v0) : "m" (*v1)
	);
	
	v0 = &m3.x;

	asm volatile (
		"vmmul.q   M200, M000, M100\n"
		"sv.q R200, 0x00+%0\n"
		"sv.q R201, 0x10+%0\n"
		"sv.q R202, 0x20+%0\n"
		"sv.q R203, 0x30+%0\n"
		
		: "+m" (*v0)
	);
	
	printMatrix("vmmul.q 1", &m3);
	nonsense(&m3);

	asm volatile (
		"vmmul.t   M201, M000, M100\n"
		"sv.q R200, 0x00+%0\n"
		"sv.q R201, 0x10+%0\n"
		"sv.q R202, 0x20+%0\n"
		"sv.q R203, 0x30+%0\n"

		: "+m" (*v0)
		);

	printMatrix("vmmul.q 2", &m3);
	nonsense(&m3);

	asm volatile (
		"vmmul.p   M200, M000, M100\n"
		"vmmul.p   M202, M002, M102\n"
		"vmmul.p   M220, M020, M120\n"
		"vmmul.p   M222, M022, M122\n"
		"sv.q R200, 0x00+%0\n"
		"sv.q R201, 0x10+%0\n"
		"sv.q R202, 0x20+%0\n"
		"sv.q R203, 0x30+%0\n"

		: "+m" (*v0)
		);

	printMatrix("vmmul.q 3", &m3);
}

int main(int argc, char *argv[]) {
	printf("Started\n");

	resetAllMatrices();
	
	checkMatrixPerVector();
	checkTranspose();
	checkMultiplyFull();
	checkMatrixIdentity();
	checkMatrixScale();
	
	printf("Ended\n");
	return 0;
}
