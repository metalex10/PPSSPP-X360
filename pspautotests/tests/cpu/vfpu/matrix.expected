Started
vtfm3.t: -10271.000000,12362.000000,-14453.000000,0.000000
vtfm4.q: -9751.000000,11802.000000,-13853.000000,15904.000000
vhtfm3.t: -10559.000000,12682.000000,-14805.000000,0.000000
vhtfm4.q: -10258.000000,12348.000000,-14438.000000,16528.000000
non transpose: 1.000000,-2.000000,3.000000,-4.000000
non transpose: 5.000000,-6.000000,7.000000,-8.000000
non transpose: 9.000000,-10.000000,11.000000,-12.000000
non transpose: 13.000000,-14.000000,15.000000,-16.000000
transpose: 1.000000,5.000000,9.000000,13.000000
transpose: -2.000000,-6.000000,-10.000000,-14.000000
transpose: 3.000000,7.000000,11.000000,15.000000
transpose: -4.000000,-8.000000,-12.000000,-16.000000
vmmul.q 1: 1585.000000,1655.000000,1787.000000,1861.000000
vmmul.q 1: 5318.000000,5562.000000,5980.000000,6246.000000
vmmul.q 1: 10514.000000,11006.000000,11840.000000,12378.000000
vmmul.q 1: 15894.000000,16634.000000,17888.000000,18710.000000
vmmul.q 2: 1585.000000,1655.000000,1787.000000,1861.000000
vmmul.q 2: 822.000000,864.000000,898.000000,6246.000000
vmmul.q 2: 3247.000000,3415.000000,3567.000000,12378.000000
vmmul.q 2: 6481.000000,6825.000000,7141.000000,18710.000000
vmmul.q 3: 337.000000,359.000000,3948.000000,4214.000000
vmmul.q 3: 1598.000000,1698.000000,6316.000000,6738.000000
vmmul.q 3: 1248.000000,1296.000000,7892.000000,8164.000000
vmmul.q 3: 3720.000000,3864.000000,11572.000000,11972.000000
vmidt: 1.000000,0.000000,3.000000,-4.000000
vmidt: 0.000000,1.000000,3.000000,-4.000000
vmidt: 1.000000,-2.000000,1.000000,0.000000
vmidt: 1.000000,-2.000000,0.000000,1.000000
vmidt: 1.000000,0.000000,0.000000,-4.000000
vmidt: 0.000000,1.000000,0.000000,-4.000000
vmidt: 0.000000,0.000000,1.000000,-4.000000
vmidt: 1.000000,-2.000000,3.000000,-4.000000
vmidt: 1.000000,0.000000,0.000000,0.000000
vmidt: 0.000000,1.000000,0.000000,0.000000
vmidt: 0.000000,0.000000,1.000000,0.000000
vmidt: 0.000000,0.000000,0.000000,1.000000
vmidt: 1.000000,1.000000,1.000000,1.000000
vmidt: 1.000000,1.000000,1.000000,1.000000
vmidt: 1.000000,1.000000,1.000000,1.000000
vmidt: 1.000000,1.000000,1.000000,1.000000
vmidt: 0.000000,0.000000,0.000000,0.000000
vmidt: 0.000000,0.000000,0.000000,0.000000
vmidt: 0.000000,0.000000,0.000000,0.000000
vmidt: 0.000000,0.000000,0.000000,0.000000
vmscl.q: 337.000000,1685.000000,3033.000000,4381.000000
vmscl.q: -674.000000,-2022.000000,-3370.000000,-4718.000000
vmscl.q: 1011.000000,2359.000000,3707.000000,5055.000000
vmscl.q: -1348.000000,-2696.000000,-4044.000000,-5392.000000
Ended
